src/Hocon.Immutable.Tests/ApiTest.cs

```diff
@@ -0,0 +1,245 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="ApiTest.cs" company="Hocon Project">
+//     Copyright (C) 2009-2018 Lightbend Inc. <http://www.lightbend.com>
+//     Copyright (C) 2013-2018 .NET Foundation <https://github.com/akkadotnet/hocon>
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Hocon.Immutable.Tests
+{
+    public class ApiTest
+    {
+        private readonly ITestOutputHelper _output;
+
+        public ApiTest(ITestOutputHelper output)
+        {
+            _output = output;
+        }
+
+        public static IEnumerable<object[]> ObjectArrayData =>
+            new List<object[]>
+            {
+                new object[] {"a=[{a=1, b=2}, {c=3,d=4}, {e=5,f=6}, ]"},
+                new object[]{@"a=[
+  {
+    a=1,
+    b=2
+  }, 
+  {
+    c=3,
+    d=4
+  }, 
+  {
+    e=5,
+    f=6
+  },
+]"},
+                new object[]{@"a=[
+{a=1,b=2}, 
+{c=3,d=4}, 
+{e=5,f=6},]"},
+                new object[]{@"a=[
+{a=1,b=2}, 
+{c=3,d=4}, 
+{e=5,f=6}]"}
+            };
+        [Theory]
+        [MemberData(nameof(ObjectArrayData))]
+        public void CanIntelligentlyUseIndexerTypeToAccessMembers(string hocon)
+        {
+            var config = Parser.Parse(hocon).ToHoconImmutable();
+
+            Assert.Equal(1, config["a"][0]["a"].GetInt());
+            Assert.Equal(2, config["a"][0]["b"].GetInt());
+
+            Assert.Equal(3, config["a"][1]["c"].GetInt());
+            Assert.Equal(4, config["a"][1]["d"].GetInt());
+
+            Assert.Equal(5, config["a"][2]["e"].GetInt());
+            Assert.Equal(6, config["a"][2]["f"].GetInt());
+        }
+
+
+        [Fact]
+        public void CanAutomaticallyCastValuesToPrimitives()
+        {
+            var hocon = @"{
+  root : {
+    bool : true
+
+    byte : 255,
+    sbyte : -128,
+
+    short : -32768,
+    ushort : 65535,
+
+    int : -2147483648,
+    uint : 4294967295,
+
+    long : -9223372036854775808,
+    ulong : 18446744073709551615,
+
+    double : 1.23
+
+    string : foo
+    quoted-string = ""foo""
+    unquoted-string = bar
+    concat-string = foo bar
+
+    object : {
+      hasContent : false
+    }
+
+    array : [1,2,3]
+    float-array : [1.0,2.0,3.0]
+    array-concat : [[1,2] [3,4]]
+    array-single-element : [1 2 3 4]
+    array-newline-element : [
+        1
+        2
+        3
+        4
+    ]
+    array-array : [[1,2],[3,4]]
+    string-array: [""1"",""2"",""3""]
+
+    timespan: 1m
+    null : null
+  },
+
+  root_2 : 1234
+}";
+            var config = Parser.Parse(hocon).ToHoconImmutable();
+            Assert.True(config["root.bool"].GetBoolean());
+            Assert.True(config["root.bool"].Get<bool>());
+            Assert.True(config["root.bool"]);
```

Comment:
Sample use of the auto-conversion getter. `ImmutableHoconLiteral` contains implicit cast operator overrides to automatically convert the string value into other primitive types. In this case, it automatically converts the string value to `bool`